src/ggplotnim/ggplot_drawing.nim

```diff
@@ -496,37 +496,92 @@ proc drawSubDf[T](view: var Viewport, fg: FilledGeom,
-                     df, i,
-                     prevVals)
+    if fg.geom.kind != gkRibbon: # we handle the ribbon points separately below
+      p = getXY(view, df, xT, yT, fg, i, theme, xOutsideRange,
+                yOutsideRange, xMaybeString = true)
+      if viewMap.len > 0:
+        # get correct viewport if any is discrete
+        viewIdx = getView(viewMap, p, fg)
+        locView = view[viewIdx]
+      if needBinWidth:
+        # potentially move the positions according to `binPosition`
+        binWidths = calcBinWidths(df, i, fg)
+        moveBinPositions(p, binWidths, fg)
+      pos = getDrawPos(locView, viewIdx,
+                       fg,
+                       p = p,
+                       binWidths = binWidths,
+                       df, i,
+                       prevVals)
     case fg.geom.position
     of pkIdentity:
       case fg.geom.kind
       of gkLine, gkFreqPoly: linePoints.add pos
+      of gkRibbon:
+        # add yMax points to end of linePoints
+        # and add yMin points to beginning of linePoints
+        block: # namespace hygiene required maybe for template from getXY?
+          # add yMax point as next point
+          p = getXY(view, df, xT, aesyMax, fg, i, theme, xOutsideRange,
+                    yOutsideRange, xMaybeString = true)
+          if viewMap.len > 0:
+            # get correct viewport if any is discrete
+            viewIdx = getView(viewMap, p, fg)
+            locView = view[viewIdx]
+          if needBinWidth:
+            # potentially move the positions according to `binPosition`
+            binWidths = calcBinWidths(df, i, fg)
+            moveBinPositions(p, binWidths, fg)
+          pos = getDrawPos(locView, viewIdx,
+                           fg,
+                           p = p,
+                           binWidths = binWidths,
+                           df, i,
+                           prevVals)
+          linePoints.add pos
+        block:
+          # add yMin point as very first point
+          p = getXY(view, df, xT, aesyMin, fg, i, theme, xOutsideRange,
+                    yOutsideRange, xMaybeString = true)
+          if viewMap.len > 0:
+            # get correct viewport if any is discrete
+            viewIdx = getView(viewMap, p, fg)
+            locView = view[viewIdx]
+          if needBinWidth:
+            # potentially move the positions according to `binPosition`
+            binWidths = calcBinWidths(df, i, fg)
+            moveBinPositions(p, binWidths, fg)
+          pos = getDrawPos(locView, viewIdx,
+                           fg,
+                           p = p,
+                           binWidths = binWidths,
+                           df, i,
+                           prevVals)
+          linePoints.insert(pos,0)
       else: locView.draw(fg, pos, p.y, binWidths, df, i, style)
     of pkStack:
       case fg.geom.kind
-      of gkLine, gkFreqPoly: linePoints.add pos
+      of gkLine, gkFreqPoly, gkRibbon: linePoints.add pos
```

Comment:
Instead of these changes, I propose the following. We already have `readErrorBar`, which handles xMin, yMin, ... etc. for error bar plots. This is essentially all that's required. Therefore, the code in this proc can remain as is, instead of the blocks from 540 to 580 and line 584 we just write:
```nim
      of gkRibbon: linePoints.addRibbonData(locView, pos, df, i, fg)
```

where `addRibbonData` is a proc you define above:
```nim
proc addRibbonData(linePoints: var seq[Coord],
                   view: Viewport,
                   pos: Coord,
                   df: DataFrame,
                   idx: int,
                   fg: FilledGeom) =
  ## adds the error bars for the ribbon to the `linePoints`.
  ## TODO: handle stacked positions properly?
  template toC1(val: float, axKind: AxisKind): Coord1D =
    block:
      var res: Coord1D
      case axKind
      of akX:
        res = Coord1D(pos: val,
                      scale: view.xScale,
                      axis: akX,
                      kind: ukData)
      of akY:
        res = Coord1D(pos: val,
                      scale: view.yScale,
                      axis: akY,
                      kind: ukData)
      res

  let (xMin, xMax, yMin, yMax) = readErrorData(df, idx, fg)
  var posMin: Coord
  var posMax: Coord
  posMin.x = if xMin.isSome: toC1(xMin.unsafeGet, akX) else: pos.x
  posMin.y = if yMin.isSome: toC1(yMin.unsafeGet, akY) else: pos.y
  posMax.x = if xMax.isSome: toC1(xMax.unsafeGet, akX) else: pos.x
  posMax.y = if yMax.isSome: toC1(yMax.unsafeGet, akY) else: pos.y
  linePoints.add posMax
  linePoints.insert(posMin, 0)
```

This has the added benefit of also supporting ribbons for `xMin, xMax` too (and even weird combinations, which will result in some funky plots I assume), and instead of adding checks along the lines of only x *or* y related procs, I'd either add: 
- a check for either xMin, xMax / yMin, yMax to `geom_ribbon`
- just explain in the doc string for `geom_ribbon` that the user should either use this or that.

Reply:
This change requires two small changes in different files.
- in `ggplot_types.nim` you need to extend the `geomKind` field for the `FilledGeom` to include `gkRibbon` to the `gkErrorBar` case, so that `xMin` etc. fields are defined for `gkRibbon`.
- in `postprocess_scales.nim`in `fillOptFields` you have to do the same, add `gkRibbon` to the `gkErrorBar` case so that these fields are actually filled and we don't have to access the `aes` of the raw `geom`.